src/transport/webrtc/substream.rs

```diff
@@ -190,49 +194,53 @@ impl tokio::io::AsyncRead for Substream {
 
 impl tokio::io::AsyncWrite for Substream {
     fn poll_write(
-        self: Pin<&mut Self>,
+        mut self: Pin<&mut Self>,
         cx: &mut Context<'_>,
         buf: &[u8],
     ) -> Poll<Result<usize, std::io::Error>> {
         if let State::SendClosed = *self.state.lock() {
             return Poll::Ready(Err(std::io::ErrorKind::BrokenPipe.into()));
         }
 
-        // TODO: try to coalesce multiple calls to `poll_write()` into single `Event::Message`
-
-        let num_bytes = std::cmp::min(MAX_FRAME_SIZE, buf.len());
-        let future = self.tx.reserve();
-        futures::pin_mut!(future);
-
-        let permit = match futures::ready!(future.poll(cx)) {
+        match futures::ready!(self.tx.poll_reserve(cx)) {
+            Ok(()) => {}
             Err(_) => return Poll::Ready(Err(std::io::ErrorKind::BrokenPipe.into())),
-            Ok(permit) => permit,
         };
 
-        let frame = buf[..num_bytes].to_vec();
-        permit.send(Event::Message(frame));
+        if buf.len() > MAX_FRAME_SIZE {
```

Comment:
nit: Should this be checked before `self.tx.poll_reserve(cx)`, such we don't waste resources and return earlier before fetching a permit?

Reply:
Sorry, forgot to push the commit that reverted this check. Did this now and added a TODO comment with a link to #502.

Reply:
Why do you need https://github.com/paritytech/litep2p/issues/502 at all?

Reply:
> Why do you need #502 at all?

To make sure it won't be forgotten about. What do you propose instead?

Reply:
Isn't it already working, as the passing test suggests?

Reply:
> Isn't it already working, as the passing test suggests?

Sorry, I didn't read the test carefully and just assumed it didn't check that all the data made it to the other side of the channel between `Substream` and `SubstreamHandle`. But it does check that and it passes. I also just noticed that the test calls `AsyncWriteExt::write_all()` but the silent truncation you flagged is in `Substream::poll_write()`. So that's why the test passes I guess.

My knowledge of tokio async stuff is pretty limited so please just tell me what you want me to do to so we can merge this PR. 😅